### PR TITLE
VPN snooze mode

### PR DESCRIPTION
**Pull request by samsymons**
Task/Issue URL: https://app.asana.com/0/72649045549333/1207974416599035/f
Tech Design URL:
CC:

**Description**:

This PR adds VPN snooze mode.

**Steps to test this PR**:
1. Run the app on-device and activate the VPN
3. Check that the "Snooze for 2 minutes" button appears, and that when you tap it the VPN enters snooze mode
4. Check your IP via search and make sure it shows your local IP
5. Go back to the VPN UI and end snooze mode, check your IP again and make sure it matches your connected server
6. Now re-enable snooze and background the app
7. Check that the Live Activity appears, and that the Wake Up button ends snooze mode
8. Re-enable snooze one more time and background the app, then wait 2 minutes for it to end
9. Make sure that snooze mode notifies you that it has ended, and the Live Activity enters the green UI state

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo/NetworkProtectionStatusViewModel.swift [hunk 1]
```diff
@@ -398,23 +440,65 @@ final class NetworkProtectionStatusViewModel: ObservableObject {
...
 
-    private class func titleText(connected isConnected: Bool) -> String {
-        isConnected ? UserText.netPStatusHeaderTitleOn : UserText.netPStatusHeaderTitleOff
+    @MainActor
+    func startSnooze() async {
+        guard let activeSession = await tunnelController.activeSession() else {
+            return
+        }
+
+        let defaultDuration: TimeInterval = .minutes(2) + .seconds(1) // TODO: Change to 20 mins, 2 min is only used for testing
```
Comment (samsymons) [inline DuckDuckGo/NetworkProtectionStatusViewModel.swift]: I'll change this before merging, 2 mins makes it easier to test expiration.

File: Widgets/Widgets.swift [hunk 2]
```diff
@@ -249,6 +249,17 @@ struct Widgets: WidgetBundle {
...
 
+struct VPNBundle: WidgetBundle {
+    @WidgetBundleBuilder
+    var body: some Widget {
+        if #available(iOS 17, *) {
+            VPNStatusWidget()
+            VPNSnoozeLiveActivity()
+        }
+    }
+}
```
Comment (samsymons) [inline Widgets/Widgets.swift:260]: The widget builder syntax only allows up to 10 elements, and the addition of the Live Activity brings up to 11, so I'm working around this with a widget bundle.

File: DuckDuckGo/NetworkProtectionStatusViewModel.swift [hunk 3]
```diff
@@ -264,12 +301,13 @@ final class NetworkProtectionStatusViewModel: ObservableObject {
...
             }
             .receive(on: DispatchQueue.main)
-            .assign(to: \.shouldShowConnectionDetails, onWeaklyHeld: self)
+            .assign(to: \.hasServerInfo, onWeaklyHeld: self)
             .store(in: &cancellables)
     }
 
     private func updateViewModel(withStatus connectionStatus: ConnectionStatus) {
-        self.headerTitle = Self.titleText(connected: connectionStatus.isConnected)
+        print("DEBUG: Updating view model with status \(connectionStatus)")
```
Comment (quanganhdo) [inline DuckDuckGo/NetworkProtectionStatusViewModel.swift]:
```suggestion
```

Comment (samsymons) [reply DuckDuckGo/NetworkProtectionStatusViewModel.swift]: Hah, thanks - this is a remnant from struggling against a weird publisher timing issue.

**Review by quanganhdo — COMMENTED** [review]
I managed to break the feature by repeatedly tapping on Snooze/Wake Up. The VPN then stuck in some state and can no longer be manipulated from the screen.

**Comment by samsymons** [comment]
@quanganhdo Thanks, I'll take a look. This PR uses a piece of state to track when a snooze request is pending, and only resets after the snooze mode has changed, but apparently it doesn't work well enough.

File: DuckDuckGo/NetworkProtectionStatusViewModel.swift [hunk 4]
```diff
@@ -434,6 +519,17 @@ final class NetworkProtectionStatusViewModel: ObservableObject {
         return UserText.netPStatusConnected(since: timeLapsed)
     }
 
+    private static func snoozeDurationRemainingMessage(for snoozeEndDate: Date, currentDate: Date = Date()) -> String {
+        if snoozeEndDate <= currentDate {
+            return UserText.netPCellSnoozing
+        }
+
+        let timeRemainingInterval = snoozeEndDate.timeIntervalSince(currentDate)
+        let timeRemaining = Self.snoozeRemainingDateFormatter.string(from: timeRemainingInterval) ?? "0:00"
```
Comment (quanganhdo) [inline DuckDuckGo/NetworkProtectionStatusViewModel.swift]: Nit

```suggestion
        let timeRemaining = Self.snoozeRemainingDateFormatter.string(from: timeRemainingInterval) ?? "00:00"
```

**Comment by quanganhdo** [comment]
The expanded live activity view doesn't seem to work properly on iOS 18.1 beta. For a split second it looks like the views are offscreen somewhere. 

Live activities for other apps work fine. 

https://github.com/user-attachments/assets/a01898ee-4ff5-4335-a187-09f31b3dcd87

**Comment by samsymons** [comment]
@quanganhdo I've fixed the issue with toggling snooze too fast. The issue was because the VPN extension was updating the snooze timing before fully starting/stopping snooze, so the app would think that snooze had changed before it really had. I can now hammer on snooze and it works great every time, but please let me know if you're still able to reproduce the issue.

File: Widgets/VPNWidget.swift [hunk 5]
```diff
@@ -141,13 +150,39 @@ struct VPNStatusView: View {
                     .fontWeight(.semibold)
                     .foregroundStyle(Color(designSystemColor: .textPrimary))
 
-                Text(status == .connected ? entry.location : UserText.vpnWidgetDisconnectedSubtitle)
-                    .font(.system(size: 12, weight: .regular))
-                    .foregroundStyle(Color(designSystemColor: .textSecondary))
-                    .opacity(status.isConnected ? 0.8 : 0.6)
+                if status == .connected {
+                    Text(snoozeTimingStore.isSnoozing ? "Until \(snoozeEndDateString)" : entry.location)
```
Comment (samsymons) [inline Widgets/VPNWidget.swift]: This needs to go into UserText.

File: Widgets/VPNWidget.swift [hunk 6]
```diff
@@ -141,13 +150,39 @@ struct VPNStatusView: View {
...
+                    Text(UserText.vpnWidgetDisconnectedSubtitle)
+                        .font(.system(size: 12, weight: .regular))
+                        .foregroundStyle(Color(designSystemColor: .textSecondary))
+                        .opacity(status.isConnected ? 0.8 : 0.6)
+                }
 
                 switch status {
-                case .connected, .connecting, .reasserting:
+                case .connected:
+                    let buttonTitle = snoozeTimingStore.isSnoozing ? "Wake Up" : UserText.vpnWidgetDisconnectButton
```
Comment (samsymons) [inline Widgets/VPNWidget.swift]: This needs to go into UserText.

File: DuckDuckGo/NetworkProtectionStatusView.swift [hunk 7]
```diff
@@ -37,7 +37,7 @@ struct NetworkProtectionStatusView: View {
             toggle()
             locationDetails()
 
-            if statusModel.isNetPEnabled && statusModel.shouldShowConnectionDetails && statusModel.ipAddress != nil {
+            if statusModel.isNetPEnabled && statusModel.hasServerInfo && !statusModel.isSnoozing {
```
Comment (quanganhdo) [inline DuckDuckGo/NetworkProtectionStatusView.swift:40]: The `ipAddress != nil` check isn't carried over to the `hasServerInfo` implementation. Is it intended?

File: DuckDuckGo/NetworkProtectionStatusView.swift [hunk 8]
```diff
@@ -125,10 +128,31 @@ struct NetworkProtectionStatusView: View {
         }
     }
 
+    @ViewBuilder
```
Comment (quanganhdo) [inline DuckDuckGo/NetworkProtectionStatusView.swift:131]: TIL we don't need to return an empty view in the else case when using @ViewBuilder.

**Review by quanganhdo — APPROVED** [review]

File: DuckDuckGo/NetworkProtectionStatusView.swift (hunk 7 above)
Comment (samsymons) [reply DuckDuckGo/NetworkProtectionStatusView.swift:40]: Yep, it was redundant since the server info includes the IP address.

**Pull request merged**